File: assets/backing-services/go/users_oidc_keycloak.go
// Package userskeycloak is a PLACEHOLDER — it is NOT a working token validator.
//
// Selecting Keycloak for the product's own users wires a second realm, customers, into the same
// Keycloak container as the staff realm app, plus the environment variables that name it. The
// token validation is deliberately not written here.
//
// The browser app (apps/web) performs the login — Authorization Code flow with PKCE, through
// react-oidc-context — and sends the access token as a bearer token. This service is a resource
// server for the customers realm: it validates that token on every request before trusting a
// byte of it. Before implementing, load skills/secure-oauth-oidc/SKILL.md — it covers what this
// package must get right and what silently breaks if it does not:
//
//   - Signature against the issuer's JWKS, found through the discovery document at
//     USERS_OIDC_ISSUER.
//   - exp: an expired token is one somebody may have lost.
//   - aud containing USERS_OIDC_AUDIENCE, so a token minted for another client is not accepted
//     here.
//   - iss exactly USERS_OIDC_ISSUER. See the mix-up hazard below.
//
// Use a maintained JWT/JWKS library rather than hand-rolling these checks.
//
// What this package owns: turning a validated token's claims into a Customer the domain can key
// on, in CustomerFromClaims. It is pure, so it is testable without a provider, and it is the one
// place these rules live:
//
//   - iss must be exactly the expected issuer.
//   - sub must be present. It is the only stable key a customer has; email addresses change.
//   - email_verified must be exactly true, because an unverified address is one anybody can type.
//
// What this package does not own: customers have no groups and no roles. Whether a customer may
// see an order is a question of ownership or tenancy, and it is decided in the use case that
// loads the order, against the Customer it is handed — not here, and not by anything in the
// token.
//
// The mix-up hazard worth knowing about now: both realms live in one Keycloak, so a staff token
// is a perfectly valid JWT, signed by the same server, whose issuer differs from a customer's by
// one path segment: /realms/app against /realms/customers. A resource server that checks the
// signature but not the issuer would accept a staff member as a customer. CustomerFromClaims
// refuses any issuer that is not exactly the expected one, and its test uses the staff realm's
// issuer on purpose. A staff token must never pass as a customer.
//
// The realm fixture is docker/keycloak/realms/customers.json; the staff one is
// docker/keycloak/realms/app.json. Locally the issuer is http://localhost:8081/realms/customers
// and the audience api; both are set in .env.example.
package userskeycloak

import (
	"errors"
	"fmt"
	"strings"
)

// Claims are the claims of a token that has already been validated. Untyped values, because a
// token carries what it likes.
type Claims map[string]any

// Customer is who the caller is, once a real validation has established it. Subject is the key;
// Email is for showing and for contacting.
type Customer struct {
	Subject string
	Email   string
}

// The reasons CustomerFromClaims refuses, so a transport can tell them apart with errors.Is.
var (
	ErrWrongIssuer     = errors.New("token issuer is not the customers realm")
	ErrNoSubject       = errors.New("token has no sub claim, and sub is the only stable key a customer has")
	ErrEmailUnverified = errors.New("token does not carry a verified email address; an unverified address is one anybody can type")
)

// ErrNotImplemented is returned by the unwritten validation.
//
// A placeholder that returned a Customer would be worse than one that refuses: every use case
// downstream would pass in tests against a customer nobody authenticated.
var ErrNotImplemented = errors.New(
	"customer token validation not implemented; see the guidance in this package and the secure-oauth-oidc skill")

// CustomerFromClaims turns a validated token's claims into a Customer, or refuses. Pure, so it is
// testable without a provider. The caller has already checked signature, exp and aud; this is
// what remains.
func CustomerFromClaims(claims Claims, expectedIssuer string) (Customer, error) {
	if issuer, _ := claims["iss"].(string); issuer != expectedIssuer {
		return Customer{}, fmt.Errorf("%w: %q is not %q; the staff realm lives in the same Keycloak, "+
			"and its tokens must never pass as a customer", ErrWrongIssuer, issuer, expectedIssuer)
	}
	subject, _ := claims["sub"].(string)
	if strings.TrimSpace(subject) == "" {
		return Customer{}, ErrNoSubject
	}
	verified, _ := claims["email_verified"].(bool)
	email, _ := claims["email"].(string)
	if !verified || strings.TrimSpace(email) == "" {
		return Customer{}, ErrEmailUnverified
	}
	return Customer{Subject: subject, Email: email}, nil
}

// Authenticate is the unwritten validation: check the bearer token's signature, exp and aud, then
// hand its claims to CustomerFromClaims.
func Authenticate(bearerToken string) (Customer, error) {
	return Customer{}, ErrNotImplemented
}

File: assets/backing-services/java/users_oidc_keycloak.java
package com.example.deliverystarter.adapters.driving.http.users;

import java.util.Map;
import java.util.Objects;
import org.jspecify.annotations.Nullable;

/**
 * Who a customer is, read from the claims of a token the framework has already validated — the part of
 * customer identity this project actually owns.
 *
 * <p><strong>The protocol flow is not here, and must never be.</strong> The browser app performs the
 * Authorization Code flow with PKCE against the {@code customers} realm and sends the access token as a
 * bearer token; the framework ({@code quarkus-oidc}, or Spring Security's resource server) retrieves the
 * realm's JWKS and validates the token's signature, {@code iss}, {@code aud} and {@code exp}. Hand-rolling
 * any of that is a security defect rather than a style choice, which is why it is nobody's code in this
 * repository. It is configured in {@code application.properties} and off until a route needs a customer;
 * read {@code skills/secure-oauth-oidc/SKILL.md} before turning it on.
 *
 * <p>What is left over is turning a validated token into something the domain can key on, and three rules
 * the framework's validation does not make for you:
 *
 * <ul>
 *   <li><strong>The issuer is the customers realm, exactly.</strong> Staff and customers are two realms in
 *       one Keycloak, so a staff token is a perfectly valid JWT from the same host whose issuer differs from
 *       a customer's by one path segment. The framework checks the issuer too; this check is here so the
 *       guarantee lives in code this project owns and tests, and survives the day someone points both
 *       realms at one tenant or one shared JWKS in configuration.
 *   <li><strong>There is a subject.</strong> {@code sub} is the identifier Keycloak keeps stable for the
 *       life of the account. An email address is something a customer can change; the domain keys on the
 *       subject, never the address.
 *   <li><strong>The email is verified, and the flag is the boolean {@code true}.</strong> An unverified
 *       address is whatever the customer typed at sign-up, and a claim that is the string {@code "true"} is
 *       a realm exporting something other than what this project expects — both are refused rather than
 *       coerced.
 * </ul>
 *
 * <p>No groups and no roles: customers are all one kind of principal. Whether <em>this</em> customer may see
 * <em>that</em> order is ownership, and ownership is decided inside use cases — a rule enforced in a
 * driving adapter is a rule the next entry point will not enforce.
 *
 * <p>Where this gets applied is the framework's own hook ({@code CurrentCustomer} on Quarkus,
 * {@code CustomerSecurityConfig} on Spring). Keeping the rules pure and the wiring separate is what lets
 * every one of them be tested with no provider running.
 *
 * @param subject the realm's stable identifier for the account, from {@code sub}
 * @param email the verified address, from {@code email}
 */
public record CustomerIdentity(String subject, String email) {

    /** The claim names read by {@link #fromClaims}, as one place to look when a realm renames one. */
    static final String ISSUER = "iss";
    static final String SUBJECT = "sub";
    static final String EMAIL = "email";
    static final String EMAIL_VERIFIED = "email_verified";

    /**
     * @throws IllegalArgumentException when either value is blank; the factory below is where a token gets
     *     to this constructor, and it has already said which rule failed
     */
    public CustomerIdentity {
        if (subject.isBlank()) {
            throw new IllegalArgumentException("a customer needs a non-blank subject");
        }
        if (email.isBlank()) {
            throw new IllegalArgumentException("a customer needs a non-blank email");
        }
    }

    /**
     * The customer a validated token's claims describe, or a refusal saying which rule failed.
     *
     * <p>Pure, so it is testable without a provider: the claims arrive as the plain map every framework can
     * produce from its own token type, and the expected issuer arrives as the configured string rather than
     * being looked up here.
     *
     * @param claims the validated token's claims, as the framework exposes them
     * @param expectedIssuer the customers realm's issuer, exactly as configured
     * @throws IllegalArgumentException naming the rule a claim broke — the issuer is another realm's, the
     *     subject is missing, the email is unverified, or the verified flag has no address to verify
     */
    public static CustomerIdentity fromClaims(Map<String, Object> claims, String expectedIssuer) {
        String issuer = Objects.toString(claims.get(ISSUER), "");
        if (!expectedIssuer.equals(issuer)) {
            throw new IllegalArgumentException(
                    "token issuer is '" + issuer + "', not the customers realm '" + expectedIssuer
                            + "'; a token from another realm in the same Keycloak is not a customer");
        }
        String subject = Objects.toString(claims.get(SUBJECT), "");
        if (subject.isBlank()) {
            throw new IllegalArgumentException(
                    "token has no subject (sub), so there is no stable identity to key on");
        }
        Object verified = claims.get(EMAIL_VERIFIED);
        if (!Boolean.TRUE.equals(verified)) {
            throw new IllegalArgumentException(
                    "email_verified is " + describe(verified) + ", not the boolean true; an unverified "
                            + "address is whatever the customer typed at sign-up");
        }
        String email = Objects.toString(claims.get(EMAIL), "");
        if (email.isBlank()) {
            throw new IllegalArgumentException(
                    "email_verified is true but the token carries no email; the realm's client scopes are "
                            + "not releasing the address");
        }
        return new CustomerIdentity(subject, email);
    }

    private static String describe(@Nullable Object value) {
        if (value == null) {
            return "absent";
        }
        return value instanceof String ? "the string \"" + value + "\"" : String.valueOf(value);
    }
}
